### PR TITLE
Improve resource installation and lookup

### PR DESCRIPTION
**Pull request by kcgen**
Previously, the resources weren't installed as part of a "meson install" step, which is used by package maintainers. 

This will now install the resources based on the specific prefix (see commits), and also look them up using the system's XDG data paths.

These details are added to the `docs/README.resources`.

### CODE REVIEW

**Comment by kcgen** [comment]
@rderooy , if you can test this - curious if it works better for your needs.

**Comment by eli-schwartz** [comment]
> Previously, the resources weren't installed as part of a "meson install" step, which is used by package maintainers.

I'm confused, shouldn't this be handled by https://github.com/dosbox-staging/dosbox-staging/blob/9e74b9f1fd4d7e6acaef666256df08a3fadf88b0/contrib/resources/meson.build#L87-L99 ?

**Comment by rderooy** [comment]
> @rderooy , if you can test this - curious if it works better for your needs.

@kcgen I can confirm that with this patch, I can build and run a flatpak without the need to copy the resource files. It is now loading resources from `/app/share/dosbox-staging` e.g.,
```
LANG: Loaded language file: /app/share/dosbox-staging/translations/de.lng
```

File: scripts/install_resources.sh
Comment (johnnovak) [reply scripts/install_resources.sh:5]: I think this is slightly easier to automatically update via scripts, I don't personally mind it and I've seen this format in other files.

File: src/misc/support.cpp
Comment (johnnovak) [inline src/misc/support.cpp:228]: 🏅

File: src/misc/support.cpp
Comment (johnnovak) [reply src/misc/support.cpp:255]: +1

**Review by johnnovak — APPROVED** [review]
Looking good @kcgen 👍🏻

**Comment by kcgen** [comment]
@eli-schwartz , yes - I've switched to putting that to use in the new revision in place of the script (just had to adjust it a little to maintain the subdirectory tree).  Thanks for pointing out I already had most of it there :sweat_smile: .

**Comment by kcgen** [comment]
Thanks for the tests, @rderooy!

If you can give this branch another try, I've updated it a bit more, and should be even more flexible with either `--prefix` and/or `--datadir` paths 

When it comes to the --datadir path, this is where you can redirect the resources to something like `/var/share` (provided one of your XGD_DATA_* paths includes it), and have the program installed to something like `--prefix /usr` (and it can be held read-only).

These are detailed in the updated readme, if you can review those changes:

https://github.com/dosbox-staging/dosbox-staging/pull/1950/commits/a4d2fdb666ef288826e7c72a452e528ce882aba2#diff-3e721f7084fd733b35083bae1fc619c1532015e8aebbab6eba50cca469fd42f5

File: docs/README.resources
Comment (kcgen) [reply docs/README.resources]: fixed, thanks!

File: scripts/install_resources.sh
Comment (kcgen) [reply scripts/install_resources.sh]: Yes, the copyright updater script also handles this "firstrevision-lastrevision" pattern, and will update the second year to the last year the file was revised.

File: src/misc/support.cpp
Comment (kcgen) [reply src/misc/support.cpp]: Deduped! https://github.com/dosbox-staging/dosbox-staging/pull/1950/commits/5fb59818288dc47a3fda5d66a99030325dba6f90

**Comment by kcgen** [comment]
Ready for round 2 review.

File: src/misc/support.cpp
Comment (Wengier) [inline src/misc/support.cpp:249]: :thumbsup:

File: src/platform/visualc/config.h
Comment (Wengier) [inline src/platform/visualc/config.h:2]: Perhaps define both `DOSBox Staging` and `dosbox-staging` here, since both are considered (official) project names? IMO `DOSBox Staging` is the usual name whereas `dosbox-staging` is more likely the internal project name.

File: src/shell/shell.cpp
Comment (Wengier) [inline src/shell/shell.cpp:1095]: Suggest displaying `DOSBox Staging` as the project name here instead of `dosbox-staging`, since the former is intended to be displayed to users.

**Comment by rderooy** [comment]
> Thanks for the tests, @rderooy!
> 
> If you can give this branch another try, I've updated it a bit more, and should be even more flexible with either `--prefix` and/or `--datadir` paths.
> 
> The `--datadir` path will let you redirect the resources to something like `/var/share` (provided one of your `XDG_DATA_*` paths includes it), while sending the program files into say, `--prefix /usr`, which can be held read-only.
> 
> These are detailed in the updated README.resources, if you can review those changes:
> 
> [a4d2fdb#diff-3e721f7084fd733b35083bae1fc619c1532015e8aebbab6eba50cca469fd42f5](https://github.com/dosbox-staging/dosbox-staging/commit/a4d2fdb666ef288826e7c72a452e528ce882aba2#diff-3e721f7084fd733b35083bae1fc619c1532015e8aebbab6eba50cca469fd42f5)

Thanks @kcgen . Will give it a try this evening.

File: src/platform/visualc/config.h
Comment (kcgen) [reply src/platform/visualc/config.h:2]: > whereas dosbox-staging is more likely the internal name used by the project (when spaces etc are not desired).

yes, that's why I named it `CANONICAL_PROJECT_NAME`, leaving the public name, `DOSBOX Staging` available for a subsequent define, perhaps just `PROJECT_NAME`.

File: src/gui/sdlmain.cpp
Comment (Wengier) [inline src/gui/sdlmain.cpp:4535]: Same as above (to be displayed as `DOSBox Staging`). Probably some other places too as long as they are displayed to users.

File: src/platform/visualc/config.h
Comment (kcgen) [reply src/platform/visualc/config.h:2]: > Perhaps define both `DOSBox Staging`

There are many more places where this is used - and this also makes up larger nouns with the same pattern, like "The DOSBox Staging Team" (as credited in the codec information strings).  In some cases, I'm reluctant to scrub our name because it's less of a branding string (which forks should replace), and actually attributing the work to the team (we wouldn't want the fork's project name to be swapped, in some cases).

This is treading a bit further than the scope of the PR. I'm a bit reluctant to expand this further (I'm going to have backport this to 0.79.1 as well.. 

But this is good stuff for a follow on PR.

File: src/shell/shell.cpp
Comment (kcgen) [reply src/shell/shell.cpp:1095]: For now, I'd like to just keep this PR with a 1:1 replacement (without these project names changing in form or how we use them). I think it's good - but should be done in it's own PR so the team can think more about them.

File: src/gui/sdlmain.cpp
Comment (kcgen) [reply src/gui/sdlmain.cpp:4535]: Many unix programs use the lower-case and string-parseable name in their version record - but will also use their public name in a verbose block.

This is usually because scripts and scanner will mass-parse all program `--version` results, and create inventories and such.  This can certainly be the case, so I'm reluctant to change this (even aside from the scope concern above).

Comment (kcgen) [reply src/gui/sdlmain.cpp:4535]: For example, GCC's official name is `GCC, the GNU Compiler Collection`, however the last line of `gcc --version`:

`gcc version 12.1.0 (Ubuntu 12.1.0-2ubuntu1~22.04) `

File: src/platform/visualc/config.h
Comment (Wengier) [reply src/platform/visualc/config.h:2]: @kcgen I did not know you decided to define CANONICAL_PROJECT_NAME here as part of this PR. So apparently the PR scope is extended with this change of yours. If you are worried about the PR scope, perhaps you can consider undo this define too, and make a follow PR for such changes.

File: src/shell/shell.cpp
Comment (Wengier) [reply src/shell/shell.cpp:1095]: If there is concern for the PR scope, perhaps you can consider undo such changes too.

File: src/gui/sdlmain.cpp
Comment (Wengier) [reply src/gui/sdlmain.cpp:4535]: The original suggestion was made in consistent with previous efforts such as https://github.com/dosbox-staging/dosbox-staging/issues/1803

If there are other concerns (PR scope etc), perhaps they can be kept as is for the purpose of this PR.

File: src/platform/visualc/config.h
Comment (kcgen) [reply src/platform/visualc/config.h:2]: This PR included duplicate uses of the `"dosbox-staging"` string, which you commented would be better if these strings were de-duped.

So I went all the way back to the source - which is Meson's metadata, where `"dosbox-staging"` is actually defines - and I used that.

The PR as-is applies an exact 1:1 de-duping if the `"dosbox-staging"` string (and as such, takes care of the original concern you raised - which was a good concern).

I'm not one to just do a partial measure (taking more effort, and having mixed results), when I can easily do it right with almost the same extra effort.

I would prefer this stand as is, and we carry on.

File: src/shell/shell.cpp
Comment (kcgen) [reply src/shell/shell.cpp:1095]: I've only de-duped the `"dosbox-staging"` string, as you commented. That's in scope - because this PR was the cause of some of that duping in the first place.

But I am reluctant to add more top-level strings for "DOSBox Staging" and further de-dupe those (because I believe doing so is more nuanced), and should have it's own PR.

File: src/platform/visualc/config.h
Comment (Wengier) [reply src/platform/visualc/config.h:2]: In the earlier Windows installer PR, you suggested to change names such as “dosbox-staging” and “DOSBox Staging” etc in the script, and you appeared to have agreed that the change can be done locally (in the script itself). I really did not know if you would go this far with this PR on the other hand.

File: src/shell/shell.cpp
Comment (kcgen) [reply src/shell/shell.cpp:1095]: To be clear, all of the commits thus-far are what I consider in-scope.

Comment (Wengier) [reply src/shell/shell.cpp:1095]: Well, you appeared to have enlarged the scope of the PR, which can be enlarged even further and further of course. I really wonder if there is a threshold for this.

Comment (kcgen) [reply src/shell/shell.cpp:1095]: I've covered the items you requested against the original code. Suggest we wrap up the PR once @rderooy is able to test.

Comment (kcgen) [reply src/shell/shell.cpp:1095]: @Wengier - I will hold off on merging the PR until you've approved; please shout if you see any remaining improvements to the current code in the PR.

**Comment by rderooy** [comment]
I just did another test with the same manifest, and the patch works. `--datadir=` is also respected when I use it.

**Comment by kcgen** [comment]
That's great news - thanks for testing both approaches, @rderooy!

File: src/shell/shell.cpp
Comment (kcgen) [reply src/shell/shell.cpp:1095]: Wengier mentioned: "IMO it can be expanded to more files too, including the Windows installer script. Ideally the define for standard project name will be added too as part of cleaning up surrounding code, and IMO they can be included in the current PR too given the config.h change"

Replying: I tried this (Using Meson to populate template values in to Windows Inno setup script), and even though it's possible, it felt miss-aligned because Meson isn't a guaranteed upstream task of the Inno Installer.

It's possible to run the installer for the VisualStudio compiler results too.

Eventually I want to move to a templating system (likely Jinja), that holds all our metadata records in a top-level metadata file (project name(s), version(s), URLs, release vs. alpha, and so on).  It will be agnostic of and run prior-to the build systems, and will be fully cross platform.  This is where we can do templating into the various packaging systems, regardless of the build system being used.

**Comment by kcgen** [comment]
Thanks for the all help with this one, @eli-schwartz , @rderooy , @Wengier , and @johnnovak.

This is in the queue for 0.79.1 too, to help any packagers who need help copying the `resources/` tree to their desired install location.

**Pull request merged**
